File: programmers/카카오 다시풀기/level2/오픈채팅방.cpp
/*
    공백으로 나누어져있어 stringstream을 사용하면 쉽게 문자열을 분해할 수 있다고 생각했다.
    map을 사용해 uid - nickname을 연결시켜주고
    다시 반복문을 돌면서 들어오고 나가고를 출력해줬다.(똑같은 반복문, stringstream을 2번 쓴게 좀 찝찝하긴 하다.)
*/
#include <string>
#include <vector>
#include <map>
#include <sstream>

using namespace std;

vector<string> solution(vector<string> record) {
    vector<string> answer;
    map<string, string> user;
    
    for(int i = 0; i < record.size(); i++) {
        
        stringstream ss(record[i]);
        string command, uid, nickname;
        
        ss >> command >> uid >> nickname;
        
        string ret = "";
        if(command[0] == 'E' || command[0] == 'C') user[uid] = nickname;
    }
    
    for(int i = 0; i < record.size(); i++) {
        
        stringstream ss(record[i]);
        string command, uid, nickname;
        
        ss >> command >> uid >> nickname;
        
        string ret = "";
        if(command[0] == 'E') {
            ret += user[uid];
            ret += "님이 들어왔습니다.";
            answer.push_back(ret);
        } else if(command[0] == 'L') {
            ret += user[uid];
            ret += "님이 나갔습니다.";
            answer.push_back(ret);
        }
    }
    
    return answer;
}

File: kakao_java/blind2019/오픈채팅방.java
package blind2019;

import java.util.*;

public class 오픈채팅방 {
    private Map<String, String> map;

    private StringTokenizer st;

    public String[] solution(String[] records) {
        map = new HashMap<>();
        for (String record : records) {
            st = new StringTokenizer(record);

            String command = st.nextToken();
            String id = st.nextToken();

            if ("Enter".equals(command)) {
                String nickName = st.nextToken();
                map.put(id, nickName);
            } else if ("Change".equals(command)) {
                String nickName = st.nextToken();
                map.put(id, nickName);
            }
        }

        List<String> result = new ArrayList<>();
        for (String record : records) {
            st = new StringTokenizer(record);

            String command = st.nextToken();
            String id = st.nextToken();

            if ("Enter".equals(command)) {
                result.add(map.get(id) + "님이 들어왔습니다.");
            } else if ("Leave".equals(command)) {
                result.add(map.get(id) + "님이 나갔습니다.");
            }
        }

        String[] answer = new String[result.size()];
        for (int i = 0; i < result.size(); i++) {
            answer[i] = result.get(i);
        }
        return answer;
    }
}
